ui/src/app/components/add-index-form/add-index-form.component.spec.ts

```diff
@@ -1,36 +1,192 @@
-
+import { of } from 'rxjs';
+import IRule from 'src/app/model/rule';
+import { ConversionService } from 'src/app/services/conversion/conversion.service';
+import { DataService } from 'src/app/services/data/data.service';
+import { SidenavService } from 'src/app/services/sidenav/sidenav.service';
+import mockIConv from 'src/mocks/conv';
 import { AddIndexFormComponent } from './add-index-form.component';
 
 describe('AddIndexFormComponent', () => {
   let component: AddIndexFormComponent;
   let fixture: ComponentFixture<AddIndexFormComponent>;
+  let dataServiceSpy: jasmine.SpyObj<DataService>;
+  let sidenavServiceSpy: jasmine.SpyObj<SidenavService>;
+  let fb: FormBuilder;
+  let conversionServiceSpy: jasmine.SpyObj<ConversionService>;
 
   beforeEach(async () => {
+    dataServiceSpy = jasmine.createSpyObj('DataService', ['conv', 'applyRule', 'dropRule']);
+    sidenavServiceSpy = jasmine.createSpyObj('SidenavService', [
+      'sidenavAddIndexTable',
+      'setSidenavAddIndexTable',
+      'closeSidenav',
+      'displayRuleFlag',
+      'ruleData',
+    ]);
+    conversionServiceSpy = jasmine.createSpyObj('ConversionService', [
+      'getTableIdFromSpName',
+      'getColIdFromSpannerColName',
+    ]);
+
     await TestBed.configureTestingModule({
-      declarations: [ AddIndexFormComponent ],
-      imports: [ReactiveFormsModule, HttpClientModule, MatSnackBarModule, MatSelectModule, BrowserAnimationsModule, MatFormFieldModule, MatInputModule]
+      declarations: [AddIndexFormComponent],
+      imports: [ReactiveFormsModule, HttpClientModule, MatSnackBarModule, MatSelectModule, BrowserAnimationsModule, MatFormFieldModule, MatInputModule],
+      providers: [
+        FormBuilder,
+        { provide: DataService, useValue: dataServiceSpy },
+        { provide: SidenavService, useValue: sidenavServiceSpy },
+        { provide: ConversionService, useValue: conversionServiceSpy },
+      ],
     })
-    .compileComponents();
+      .compileComponents();
+
+    fb = TestBed.inject(FormBuilder);
+
+    conversionServiceSpy.getColIdFromSpannerColName.and.returnValue("TestId")
+    conversionServiceSpy.getTableIdFromSpName.and.returnValue("t1")
+    dataServiceSpy.conv = of(mockIConv);
+    dataServiceSpy.dropRule.and.callThrough();
+    sidenavServiceSpy.sidenavAddIndexTable = of("t1");
+    sidenavServiceSpy.displayRuleFlag = of(true)
+    sidenavServiceSpy.ruleData = of()
   });
 
   beforeEach(() => {
     fixture = TestBed.createComponent(AddIndexFormComponent);
     component = fixture.componentInstance;
+    component.addIndexForm = fb.group({
+      tableName: [''],
+      indexName: ['', []],
+      ColsArray: fb.array([]),
+    });
     fixture.detectChanges();
   });
 
   it('should create', () => {
     expect(component).toBeTruthy();
   });
+
+  it('should render ngOnInit and all dependent services correctly', () => {
+    const addIndexRule: IRule = {
```

Comment:
This is returned from the `sidenavServiceSpy` mock in response to a call to that service, it should also be asserted somewhere.